### ISSUE TITLE
External parsed entities in the internal DTD subset of an SVG content document

### ISSUE THREAD

**Issue by murata2makoto** [w3c/epub-specs#1355]
In my understanding, EPUB 3.2 does not allow this XML document as an SVG content document.

Here `desc `is an external parsed entity declared in an internal DTD subset.   The content of 
desc.ent is "&lt;desc>&lt;/desc>", possibly preceded by the XML declaration.

Do people agree that this is not an SVG content document, as specified in EPUB 3.0, 3.0.1, and
3.2?   The proposed resolution (see #1354) appears to allow this as an SVG content document.  Is this intentional?

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE svg [
<!ENTITY desc SYSTEM "desc.ent">]>
<svg xmlns="http://www.w3.org/2000/svg" width="100%" height="100%">
    &desc;
    <g alignment-baseline="baseline"></g>
</svg>
```

**Comment by llemeurfr** [comment]
The solution proposed by the WG being in https://github.com/w3c/epub-specs/issues/1338#issuecomment-715919230, I agree with Makoto that the sample above is compliant with this proposal BUT causes problems to offline reading systems (especially if "desc.ent" is replaced by a full URL) and can trigger XML external entities attacks, therefore should not be compliant with the EPUB spec.

**Comment by Doktorchen** [comment]
Another example for an XHTML use case (why only SVG?).
In poetry, lyrics, recipes, bills, schedules, non-linear texts as well as in formulas (MathML, terms) one repeats content - a use case for simple entities:

<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html SYSTEM "meins.dtd">
<html xml:lang="de"
      xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>&t1;</title>
  </head>
  <body>
   <section>
    <h1 class="&c1; &c2;">&t1;</h1>
    &p1;
    &p2;
   </section>
   <section>
    <h1 class="&c3; &c2;">&t1;</h1>
    &p2;
    &p2;
   </section>
  </body>
</html>


meins.dtd is a file in the EPUB archive with simple content like this:

<!ENTITY t1 "Hallo Welt!">
<!ENTITY p1 "<p>Ein Absatz repräsentiert einen abgeschlossenen Gedankengang.</p>">
<!ENTITY p2 "<p>Ein Dokument kann natürlich viele Absätze enthalten.</p>">
<!ENTITY c1 "meins-oben">
<!ENTITY c2 "meins-beispiel">
<!ENTITY c3 "meins-unten">

Of course, 'meins.dtd' must not contain any URI/IRI references to external subsets .

Presumably such construct would expose, if some user-agent, viewer or reading systems uses an HTML5 tag soup parser instead of the required XML parser ...
Presumably the tag soup parser will run intro trouble to resolve the entities.

**Comment by Doktorchen** [comment]
Oh - the github parser seems to corrupt the example and these symbols to edit seem not to work in my browser - no accessible (unscripted) techniques available at github? ;o)

**Comment by Doktorchen** [comment]
Hopefully with entities it works better:

&lt:?xml version="1.0" encoding="UTF-8" ?&gt:
&lt:!DOCTYPE html SYSTEM "meins.dtd"&gt:
&lt:html xml:lang="de"
      xmlns="http://www.w3.org/1999/xhtml"&gt:
  &lt:head&gt:
    &lt:title&gt:&t1;&lt:/title&gt:
  &lt:/head&gt:
  &lt:body&gt:
   &lt:section&gt:
    &lt:h1 class="&c1; &c2;"&gt:&t1;&lt:/h1&gt:
    &p1;
    &p2;
   &lt:/section&gt:
   &lt:section&gt:
    &lt:h1 class="&c3; &c2;"&gt:&t1;&lt:/h1&gt:
    &p2;
    &p2;
   &lt:/section&gt:
  &lt:/body&gt:
&lt:/html&gt:


meins.dtd is a file in the EPUB archive with simple content like this:

&lt:!ENTITY t1 "Hallo Welt!"&gt:
&lt:!ENTITY p1 "&lt:p&gt:Ein Absatz repräsentiert einen abgeschlossenen Gedankengang.&lt:/p&gt:"&gt:
&lt:!ENTITY p2 "&lt:p&gt:Ein Dokument kann natürlich viele Absätze enthalten.&lt:/p&gt:"&gt:
&lt:!ENTITY c1 "meins-oben"&gt:
&lt:!ENTITY c2 "meins-beispiel"&gt:
&lt:!ENTITY c3 "meins-unten"&gt:

**Comment by murata2makoto** [comment]
@Doktorchen 

I'm trying to solve one issue at a time.  Please start another thread for your question, since it is not about SVG.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/publishing/groups/epub-wg/Meetings/Minutes/2020-11-06-epub#section3).

- `RESOLVED:  Merge PR #1368 to address outstanding DTD issues, and close GH issues 1369-1373`
<details><summary><i class="differentiate">View the transcript</i></summary>
Wendy Reid: we had resolutions at the F2F, and further discussions on github<br/>
… and came to a happy place<br/>
Matt Garrish: <a href="https://github.com/w3c/epub-specs/pull/1368">https://github.com/w3c/epub-specs/pull/1368</a><br/>
Matt Garrish: where we ended up was…<br/>
… we put in an allowance for a specific set of external identifiers that we have put in an appendix<br/>
… we have SVG and MathML that are allowed to be used in content docs or in separate files<br/>
… and we made a restriction against external entities in the internal DTD subset<br/>
… so it prevents some security issues but eases authoring<br/>
… so we’ll no longer force people to remove SVG DTDs from tool-generated files<br/>
… I’m hoping this is it :)<br/>
Ivan Herman: tech comment<br/>
… in fact, the changes are such that<br/>
… makes possible something that I’m not sure we really use<br/>
… I can define as part of an internal entity something that won’t go out to the network<br/>
… I’m not sure if this feature is in use<br/>
… formal comment<br/>
… there was a formal resolution on the previous version; this PR slightly changes that<br/>
… can we get a formal resolution to merge, and also close a bunch of issues which were examples of the problem?<br/>
Proposed resolution: Merge PR #1368 to address outstanding DTD issues, and close GH issues 1369-1373 (Wendy Reid)<br/>
Garth Conboy: +1<br/>
Matt Garrish: +1<br/>
Ivan Herman: +1<br/>
Charles LaPierre: +1<br/>
Matthew Chan: +1<br/>
Wendy Reid: +1<br/>
Brady Duga: +1<br/>
George Kerscher: +1<br/>
Laura Brady: +1<br/>
Bill Kasdorf: +1<br/>
Ben Schroeter: +1<br/>
Resolution #1: Merge PR #1368 to address outstanding DTD issues, and close GH issues 1369-1373<br/>
</details>

**Comment by iherman** [comment]
@wareid I believe this issue should be closed, too
